src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,288 @@
+}
+
+extern "C" uint32_t NetSecurityNative_InitSecContext(uint32_t* minorStatus,
+                                                     GssCredId* claimantCredHandle,
+                                                     GssCtxId** contextHandle,
+                                                     uint32_t isNtlm,
+                                                     GssName* targetName,
+                                                     uint32_t reqFlags,
+                                                     uint8_t* inputBytes,
+                                                     uint32_t inputLength,
+                                                     struct PAL_GssBuffer* outBuffer,
+                                                     uint32_t* retFlags)
+{
+    assert(minorStatus != nullptr);
+    assert(claimantCredHandle != nullptr);
+    assert(contextHandle != nullptr);
+    assert(isNtlm == 0 || isNtlm == 1);
+    assert(targetName != nullptr);
+    assert(contextHandle != nullptr);
+    assert(outBuffer != nullptr);
+    assert(retFlags != nullptr);
+
+    // inputBytes will be null here in the first call to the gss_init_sec_context but not null
+    // in subsequent calls. There's no reasonable way to assert on input
+    // other than the following check
+    if (inputLength > 0) {
+        assert(inputBytes != nullptr);
+    }
+
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID desiredMech = isNtlm ? GSS_NTLM_MECHANISM : GSS_SPNEGO_MECHANISM;
+#else
+    assert(!isNtlm && "NTLM is not supported by MIT libgssapi_krb5");
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .elements = static_cast<void*>(gss_mech_value)};
+    gss_OID desiredMech = &gss_mech_spnego_OID_desc;
+#endif
+
+    GssBuffer inputToken {.length = UnsignedCast(inputLength), .value = inputBytes};
+    GssBuffer gssBuffer { .length = 0, .value = nullptr };
+
+    uint32_t majorStatus = gss_init_sec_context(minorStatus,
+                                                claimantCredHandle,
+                                                contextHandle,
+                                                targetName,
+                                                desiredMech,
+                                                reqFlags,
+                                                0,
+                                                GSS_C_NO_CHANNEL_BINDINGS,
+                                                &inputToken,
+                                                nullptr,
+                                                &gssBuffer,
+                                                retFlags,
+                                                nullptr);
+
+    return NetSecurityNative_HandleError(majorStatus, &gssBuffer, outBuffer);
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseCred(uint32_t* minorStatus, GssCredId** credHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(credHandle != nullptr);
+
```

Comment:
Do we need to assert anything about `*credHandle`?  (Repeat for every other double-pointer parameter)